registry/client.go

```diff
@@ -0,0 +1,278 @@
+package registry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/docker/utils"
+	"github.com/go-fsnotify/fsnotify"
+)
+
+var (
+	caPool       *x509.CertPool
+	clientCerts  []tls.Certificate
+	certLock     sync.RWMutex
+	certWatcher  *fsnotify.Watcher
+	certsDirname = "/etc/docker/certs.d"
+)
+
+func init() {
+	// Setup initial certificate pool and client certificates.
+	loadCerts(watchCertDirs())
+}
+
+type TimeoutType uint32
+
+const (
+	NoTimeout TimeoutType = iota
+	ReceiveTimeout
+	ConnectTimeout
+)
+
+// Client should handle http requests for all registry sessions.
+type Client struct {
+	httpClient *http.Client
+	isSecure   bool
+}
+
+// Do handles making an http request and ensures that
+// Auth headers are never sent over an insecure channel.
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader != "" && !(c.isSecure && req.URL.Scheme == "https") {
```

Comment:
perhaps there are too many boolean operators on this line making it too complex :(

the condition in parentheses (`c.isSecure && req.URL.Scheme == "https"`) would mean, in plain english, that the request will use "real" secure TLS for the connection not `InsecureSkipVerify`. If this is not the case then the client will refuse to make a request with any Authorization header.